contributors/design-proposals/cluster-lifecycle/draft-20171020-bootstrap-checkpointing.md

```diff
@@ -0,0 +1,108 @@
+# Kubernetes Bootstrap Checkpointing Proposal
+
+## Metadata
+
+```yaml
+kep-number: draft-20171029-boostrap-checkpointing.md
+title: Kubernetes Bootstrap Checkpointing Proposal
+authors:
+  - name: Timothy St. Clair
+    github: @timothysc
+    email: timothysc@gmail.com
+owning-sig: sig-cluster-lifecycle 
+participating-sigs:
+  - sig-node
+reviewers:
+  - @yujuhong 
+  - @luxas 
+  - @roberthbailey
+approvers:
+  - @yujuhong
+  - @roberbailey
+editor:
+  name: @timothysc
+creation-date: 2017-10-20
+last-updated: 2017-10-20
+status: draft
+```
+
+## Table of Contents
+
+* [Summary](#summary)
+* [Objectives](#objectives)
+  * [Goals](#goals)
+  * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+  * [User Stories](#user-stories)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Unresolved Questions](#unresolved-questions)
+
+## Summary
+
+There are several methods to deploy a kubernetes cluster, one method that offers some unique advantages is self hosting.  The purpose of this proposal is to outline a method to checkpoint specific annotated pods, namely the control plane components, for the purpose of enabling self hosting.  
+
+The details of self hosting are beyond the scope of this proposal, and are outlined in the references listed below: 
+
+  - [Self Hosted Kubernetes][0]
+  - [Kubeadm Self Hosting][1]
+
+Extra details on this proposal, and its history, can be found in the links below: 
+
+  - [Bootstrap Checkpointing Draft 1][2]
+  - [Bootstrap Checkpointing Draft 2][3]
+  - [WIP Implementation][4]
+
+## Objectives
+
+The scope of this proposal is **bounded**, but has the potential for broader reuse in the future.  The reader should be mindful of the explicitly stated [Non-Goals](#non-goals) that are listed below.
+
+### Goals
+ 
+ - Provide a basic framework for recording annotated *Pods* to the filesystem. 
+ - Ensure that a restart of the kubelet checks for existence of these files and loads them on startup.
+
+### Non-Goals
+
+- This is not a generic checkpointing mechanism for arbitrary resources. (e.g. Secrets)  Such changes require wider discussions.
+- This will not checkpoint internal kubelet state. 
+- This proposal does not cover self hosted kubelet(s).  It is beyond the scope of this proposal, and comes with it's own unique set of challenges.
+
+## Proposal
+The enablement of this feature is gated by a single command line flag that is passed to the kubelet on startup, ```--bootstrap-checkpoint-path``` , and will be denoted that it is experimental. 
```

Comment:
Is it common to put experimental into the flag name to denote it as such? Or does that just cause a headache when moving to a stable implementation?

@mtaufen - I'm also curious how we will denote experimental / alpha / whatever features flags in dynamic kubelet config.

Reply:
If it's experimental/alpha/whatever please feature gate it (really we should just call these things alpha, experimental is an old colloquialism...).

For managing alpha fields in any componentconfig payload, see https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#alpha-field-in-existing-api-version.

Reply:
It's not a field so that like doesn't apply.  I'm following the convention for *annotations which is to remove that (alpha) from it b/c it breaks across versioning.  This is consistent on other PRs.  

With regards to the flag naming, I'm happy to add (alpha, or experimental) but adding feature gate to symbol pollute other binaries where it only applies to the kubelet is also an anti-pattern now.  I'll take final resolution from @yujuhong , but I'm not going to symbol pollute other binaries.

Reply:
The new guideline discourages using annotations for alpha/beta features. I think one important question is that do we plan to convert the annotation (`node.kubernetes.io/bootstrap-checkpoint=true`) to a regular field in the future when it becomes GA? If so, we should follow the guideline to make it an alpha field and gate the feature properly.

> adding feature gate to symbol pollute other binaries where it only applies to the kubelet is also an anti-pattern now

If there is API change, I think feature gate is generally expected. I'd be slightly less concerned if the change is restricted to a new kubelet flag.